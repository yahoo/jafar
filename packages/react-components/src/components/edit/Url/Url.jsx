/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import Input from '@material-ui/core/Input';

/**
 * Represent a string value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Url/Url.jsx">
 Url</a> from '@jafar-org/react-components/edit/Url'
 */
export default class Url extends React.Component {
  static propTypes = {
    value: PropTypes.string,
    state: PropTypes.shape({
      placeholder: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    invalid: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: '',
    state: {
      placeholder: '',
    },
    disabled: false,
    invalid: false,
  };

  render() {
    return (
      <Input
        type="url"
        value={this.props.value}
        placeholder={this.props.state.placeholder}
        disabled={this.props.disabled}
        error={this.props.invalid}
        onChange={this.onValueChange} />
    );
  }

  onValueChange = (e) => {
    this.props.onValueChange(e.target.value);
  }
}
