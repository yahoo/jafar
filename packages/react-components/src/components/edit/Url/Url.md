<h3>Examples</h3>

<h4>Usage in jafar form</h4>

 ```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

 const model = {
  id: 'simple',
  fields: {
    favoriteWebsite: {
      label: 'Favorite Website',
      path: 'favoriteWebsite',
      component: {
        name: 'Url',
        state: {
          placeholder: 'Enter url...',
        }
      },
      required: true,
      validators: [{
        name: 'url',
      }],
      disableTerm: {
        name: 'disableMe',
      },
      excludeTerm: {
        name: 'excludeMe',
      },
    },
  },
  data: {
      favoriteWebsite: 'http://test.com', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
     Url: {
        renderer: Url,
    },  
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='favoriteWebsite' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 'http://test.com',
    state: {
        placeholder: 'Enter Url',
    },
};

 <Url
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    value: undefined,
    state: {
        placeholder: 'Enter Url',
    },
};
 <Url
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Disabled</h4>

```javascript
initialState = { 
    value: 'http://test.com',
    state: {
        placeholder: 'Enter Url',
    },
};
 <Url
    value={state.value}
    state={state.state}
    disabled={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Invalid</h4>

```javascript
initialState = { 
    value: 'http://test.com',
    state: {
        placeholder: 'Enter Url',
    },
};
 <Url
    value={state.value}
    state={state.state}
    invalid={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```