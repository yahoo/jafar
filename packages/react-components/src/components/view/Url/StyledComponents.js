/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import styled from 'styled-components';

export const A = styled.a`
  color: ${props => props.color};
  cursor: pointer;
  text-decoration: none;
  display: ${props => (props.maxLines ? '-webkit-box;' : 'initial')};
  -webkit-line-clamp: ${props => (props.maxLines ? props.maxLines : 'initial')};
  -webkit-box-orient: ${props => (props.maxLines ? 'vertical' : 'initial')};
  overflow: ${props => (props.maxLines ? 'hidden' : 'initial')};
  word-break: ${props => (props.maxLines ? 'break-all' : 'initial')};
  &:hover {
    text-decoration: underline;
  }
`;

export default {
  A,
};
