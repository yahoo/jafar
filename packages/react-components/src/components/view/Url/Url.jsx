/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import { withTheme } from '@material-ui/core/styles';
import Styled from './StyledComponents';

/**
 * Represent a string value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/view/Url/Url.jsx">
 Url</a> from '@jafar-org/react-components/view/Url'
 */
class Url extends React.Component {
  static propTypes = {
    value: PropTypes.string,
    state: PropTypes.shape({
      label: PropTypes.string,
      target: PropTypes.string,
      maxLines: PropTypes.number,
    }),
  };

  static defaultProps = {
    value: '',
    state: {
      label: '',
      target: '_blank',
      maxLines: undefined,
    },
  };

  render() {
    const { theme } = this.props;
    return (<Styled.A maxLines={this.props.state.maxLines}
      color={theme.palette.secondary.main}
      href={this.props.value}
      target={this.props.state.target}>
      {this.props.state.label || this.props.value}</Styled.A>);
  }
}
export default withTheme(Url);
