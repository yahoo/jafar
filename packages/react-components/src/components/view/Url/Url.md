<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    facebookUrl: {
      label: 'Facebook Url',
      path: 'facebookUrl',
      component: {
        name: 'Url',
      },
      validators: [{
        name: 'url',
      }],
    },
  },
  data: {
    facebookUrl: 'https://www.facebook.com/friends.tv/',
  },
};

const resources = {
  components: { 
    Url: {
      renderer: Url,
    },
  }
};

<Form model={model} resources={resources}>
    <Field id='facebookUrl' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 'https://www.facebook.com/friends.tv/'
};

<Url value={state.value} />
```

<h4>With label and target</h4>

```javascript
initialState = { 
    value: 'https://www.facebook.com/friends.tv/',
    state: {
        label: 'Friends facebook page',
        target: '_self',
    }
};

<Url value={state.value} state={state.state} />
```

<h4>Max lines</h4>

```javascript
initialState = { 
    value: 'https://www.facebook.com/friends.tv/aslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefoaslkfmalkaslfksjflwfjlsfjwefijcsiefjwoiejfwoeifjweoifjwefiojweofijwefijwefo',
    state: {
        maxLines: 3,
    }
};

<Url value={state.value} state={state.state} />
```
