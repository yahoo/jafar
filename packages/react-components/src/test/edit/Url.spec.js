/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */
 
import React from 'react';
import { mount, shallow } from 'enzyme';
import Url from '../../components/edit/Url/index';

describe('<Url />', () => {
  let component;
  let onValueChangeSpy;
  const value = 'http://test.com';
  const state = { placeholder: 'Enter URL' };
  const disabled = false;
  const invalid = false;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onValueChange callback on value change', () => {
    component = mount(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    const newValue = 'http://test-new.com';
    component.find('input').simulate('change', { target: { value: newValue } });
    expect(onValueChangeSpy).toHaveBeenCalledWith(newValue);
  });

  function getComponent(value, state, disabled, invalid, onValueChange, multiline) {
    return (
      <Url
        value={value}
        state={state}
        disabled={disabled}
        invalid={invalid}
        onValueChange={onValueChange}
      />
    );
  }
});
