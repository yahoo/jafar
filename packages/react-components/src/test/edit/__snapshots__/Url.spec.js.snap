// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Url /> should render provided data 1`] = `
<WithStyles(ForwardRef(Input))
  disabled={false}
  error={false}
  onChange={[Function]}
  placeholder="Enter URL"
  type="url"
  value="http://test.com"
/>
`;
